### ISSUE TITLE
SO100 Robot Unable to Lift Objects

### ISSUE THREAD

**Issue by 2q46** [haosulab/ManiSkill#1427]
Hi, I am trying to work on parallelizing motion-planning using PyRoki with Maniskill but I am having some trouble solving some basic tasks, such as PickCube using an SO100 Robot. I created my own PickCube environment but when I attempt to grasp the cube and lift the robot is unable to lift the cube, unless I set the mass of the cube to a very small number, for which case it works. Also I am getting some issues saying "Sapien is not responding". As an example, I've attached the two videos for both scenarios below:

Cube mass set to almost zero: https://github.com/user-attachments/assets/63544996-75c1-481f-97bb-396516965c72

Cube mass set to 0.02: https://github.com/user-attachments/assets/f179f6be-1a01-4c82-b6ec-eef9f2ca0df9

Thanks in advance. Hope someone can help me resolve this issue.

**Comment by StoneT2000** [comment]
I'm not that experienced with PyRoki but one issue is that the SO100 robot is 5DOF, which makes it harder to generate motions since not all possible end-effector poses are attainable.

For grasping issue, I recommend you try the PickCube environment we have tuned already for SO100 (primarily for RL). You can do

```python
env = gym.make("PickCube-v1", robot_uids="so100")
```

**Comment by IronPlanet** [comment]
Adding reproducible evidence for this issue ("needs context").

**Finding: `is_grasped` never fires for the SO100 agent, even with the cube forced into the closed jaws — so all grasp-gated reward in `PickCubeEnv` is unreachable, and RL on `PickCubeSO100-v1` cannot learn.**

Setup: master @ 62ff3a5 (editable install), torch 2.13.0+cu130, RTX 5060 Ti 16GB, Linux.

**Probe** (cube teleported to the TCP, gripper commanded closed for 15 steps):

```python
import torch, gymnasium as gym, sapien
import mani_skill.envs  # noqa

env = gym.make("PickCubeSO100-v1", num_envs=1, obs_mode="state",
               sim_backend="physx_cuda", control_mode="pd_joint_delta_pos")
env.reset(seed=0)
base = env.unwrapped
base.cube.set_pose(sapien.Pose(base.agent.tcp_pose.p[0].cpu().numpy().tolist()))
act = torch.zeros(env.action_space.shape, device="cuda")
act[..., -1] = -1.0   # also tried +1.0
for _ in range(15):
    obs, r, term, trunc, info = env.step(act)
print(bool(info["is_grasped"][0]))   # -> False (both gripper polarities)
```

Output: `is_grasped: False` with reward ~0.13 (reaching term only), for both gripper action polarities.

**Corroborating training evidence:**
- `PickCubeSO100-v1` with the repo's own recipe (`ppo_fast.py`, 4096 envs, num-steps 8, update_epochs 8, num_minibatches 32, 50M steps): `train/success_once` and `eval/success_once` flat 0.0 for the entire run (occasional 4.5% blips), return plateaus at the reaching-reward level (~5.7).
- Control: identical recipe on `PickCube-v1` (Panda) reaches 100% eval success in 3M steps on the same machine — harness and reward code exonerated; the failure is specific to the SO100 grasp detection path.

Happy to run further diagnostics on this setup if useful.

**Comment by sai-suri** [comment]
Adding the mechanism behind @IronPlanet's probe. Their oracle test isolates the problem to the grasp path; this narrows it to the gripper drive, and there is a workaround that needs no API change.

**The two shipped SO-100 grasping envs use different gripper drives:**

```
env_id               default control_mode        use_delta use_target force_limit
PickCubeSO100-v1     pd_joint_delta_pos               True      False         100
SO100GraspCube-v1    pd_joint_target_delta_pos        True       True         100
```

`PickCubeSO100-v1` (`envs/tasks/tabletop/pick_cube.py:197`) sets `robot_uids="so100"` and inherits the default. `SO100GraspCube-v1` (`envs/tasks/digital_twins/so100_arm/grasp_cube.py`) explicitly sets the target variant. In `agents/robots/so100/so_100.py` these are the same config with `use_target` False and True respectively.

With `use_target=False` the controller recomputes its target from measured position each step, so a *held* close command is a sustained squeeze that keeps climbing rather than settling into a position hold.

**Oracle grasp, cube placed at the jaw centre, 10 seeds, 15.6 g cube, `physx_cpu`:**

```
strategy         control_mode                    held  is_grasped  peak_force_N  median_offset_mm
delta_hold       pd_joint_delta_pos              0/10        0/10          79.7             370.5
delta_release    pd_joint_delta_pos             10/10       10/10          22.0               1.1
target_hold      pd_joint_target_delta_pos       0/10        0/10         102.0              82.6
target_release   pd_joint_target_delta_pos       0/10        0/10          73.9              75.7
abs_latch        pd_joint_pos                   10/10       10/10           2.4               2.7
```

`_hold` holds the close action; `_release` commands full close until contact, then 0. Held/is_grasped counts are identical across runs; forces and offsets vary somewhat run to run.

In `delta_hold` the force climbs 0 -> 19 -> 47 -> 76 N over four steps at 20 Hz, contact then drops to zero and the cube is ~86 mm from the jaw centre by step 5. It is ejected, not dropped.

**Workaround needing no API change:** on `pd_joint_delta_pos`, command full close until contact, then command `0` (`delta_release`, 10/10 at ~22 N peak). Commanding zero stops the squeeze from accumulating and the grasp holds. Latching an absolute target at first contact (`pd_joint_pos`, 0.006 rad/step ramp) is gentler still at 2.4 N.

**Note for anyone assuming the sibling env's drive is the fix: it is not, at this closing rate.** Both target-variant strategies fail the same oracle 0/10, and `target_hold` peaks higher than the delta drive (102 N). The target ratchets past the blocked joint, so zeroing the action latches a deep squeeze rather than releasing it.

**Suggested resolution:** document the drive difference between the two envs and the zero-after-contact pattern. A default change to the target variant would not on its own fix the grasp.

Platform: mani_skill 3.0.1, macOS, `physx_cpu`, torch 2.13.0. @IronPlanet's setup was Linux/CUDA, so this is a second platform rather than a direct reproduction of their `is_grasped` result — I have not confirmed that `use_target` explains their CUDA failure specifically.

Full script: https://gist.github.com/sai-suri/58bf8e5bfe27c43cb4a8539a92ec1e43
